toolchain/check/impl_lookup.cpp

```diff
@@ -1016,62 +1150,50 @@ auto EvalLookupSingleImplWitness(Context& context, SemIR::LocId loc_id,
     auto result = GetWitnessIdForImpl(
         context, loc_id, kind == Concrete, query_self_const_id,
         query_specific_interface, candidate.impl_id);
-    if (result.has_value()) {
-      // Record the query which found a final impl witness. It's illegal to
-      // write a final impl afterward that would match the same query.
-      //
-      // If the impl was effectively final, then we don't need to poison here. A
-      // change of query result will already be diagnosed at the point where the
-      // new impl decl was written that changes the result.
-      if (mode != EvalImplLookupMode::RecheckPoisonedLookup &&
-          result.has_final_value() &&
-          !IsImplEffectivelyFinal(context,
-                                  context.impls().Get(candidate.impl_id))) {
-        context.poisoned_concrete_impl_lookup_queries().push_back(
-            {.loc_id = loc_id,
-             .query = eval_query,
-             .impl_witness = result.final_witness()});
-      }
+    if (!result.has_value()) {
+      continue;
+    }
 
-      if (kind == Concrete && candidates.consider_cpp_candidates) {
-        // We found a Carbon impl. Also check for a C++ candidate that is a
-        // better match than that impl.
-        auto cpp_witness_id = LookupCppImpl(
-            context, loc_id,
-            GetFacetAsType(context, loc_id, query_self_const_id),
-            query_specific_interface, &candidate.type_structure,
-            SemIR::LocId(
-                context.impls().Get(candidate.impl_id).first_owning_decl_id));
-        if (cpp_witness_id.has_value()) {
-          result = EvalImplLookupResult::MakeFinal(cpp_witness_id);
-        }
-      }
+    // Record the query which found a final impl witness. It's illegal to
+    // write a final impl afterward that would match the same query.
+    //
+    // If the impl was effectively final, then we don't need to poison here. A
+    // change of query result will already be diagnosed at the point where the
+    // new impl decl was written that changes the result.
+    if (use_cache && result.has_final_value() &&
```

Comment:
Can we just pass in the `mode` and use that? It's also a bool but with more meaningful names